### PR TITLE
Refs #24686 -- Made AlterField operation a noop when renaming related model with db_table.

### PR DESCRIPTION
**Pull request by DevilsAutumn**
Ref PR : #16905 
Ref Ticket : [#24686](https://code.djangoproject.com/ticket/24686)

### CODE REVIEW

File: tests/migrations/test_operations.py [hunk 1]
```diff
@@ -988,6 +988,52 @@ def test_rename_model_with_superclass_fk(self):
...
+            False,
+        )
+        self.assertIs(
+            any("DROP TABLE" in query["sql"] for query in ctx.captured_queries),
+            False,
+        )
+        self.assertIs(
+            any("ALTER TABLE" in query["sql"] for query in ctx.captured_queries),
+            False,
+        )
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1035]: We can assert that there were no queries instead.

File: tests/migrations/test_operations.py [hunk 2]
```diff
@@ -988,6 +988,52 @@ def test_rename_model_with_superclass_fk(self):
...
+
+        new_state = project_state.clone()
+        operation = migrations.RenameModel("Rider", "NewRider")
+        operation.state_forwards(app_label, new_state)
+
+        with connection.schema_editor() as editor, CaptureQueriesContext(
+            connection
+        ) as ctx:
+            operation.database_forwards(app_label, editor, project_state, new_state)
```
Comment (felixxm) [inline tests/migrations/test_operations.py:1022]:
```suggestion
        with connection.schema_editor() as editor, self.assertNumQueries(0):
            operation.database_forwards(app_label, editor, project_state, new_state)
```

File: django/db/backends/base/schema.py [hunk 3]
```diff
@@ -1601,10 +1601,19 @@ def _field_should_be_altered(self, old_field, new_field, ignore=None):
...
         # - changing an attribute in the provided set of ignored attributes
         # - adding only a db_column and the column name is not changed
+        # - db_table does not change for model referenced by FK/M2M fields
         for attr in ignore.union(old_field.non_db_attrs):
             old_kwargs.pop(attr, None)
         for attr in ignore.union(new_field.non_db_attrs):
             new_kwargs.pop(attr, None)
+        if (old_field.related_model and new_field.related_model) and not isinstance(
+            new_field.related_model, str
+        ):
```
Comment (felixxm) [inline django/db/backends/base/schema.py:1611]: I switched to `remote_field`.

**Review by felixxm — COMMENTED** [review]
@DevilsAutumn Thanks :+1: 

I pushed small edits.

File: django/db/backends/base/schema.py [hunk 4]
```diff
@@ -1601,10 +1601,19 @@ def _field_should_be_altered(self, old_field, new_field, ignore=None):
...
         for attr in ignore.union(new_field.non_db_attrs):
             new_kwargs.pop(attr, None)
+        if (
+            old_field.remote_field
+            and new_field.remote_field
+            and old_field.remote_field.model._meta.db_table
+            == new_field.remote_field.model._meta.db_table
+        ):
+            old_kwargs.pop("to", None)
+            new_kwargs.pop("to", None)
```
[the remark this answers is not in the document]
Comment (DevilsAutumn) [reply django/db/backends/base/schema.py:1617]: Yes, but while moving a model it should be ignored, right?

[the remark this answers is not in the document]
Comment (DevilsAutumn) [reply django/db/backends/base/schema.py:1617]: I was asking for the case, when a model is moved which has `db_table` and is referenced by a m2m field. In that case, AlterField should be a noop, but this patch will ignore that field too. 🤔

[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/backends/base/schema.py:1617]: > I was asking for the case, when a model is moved which has `db_table` and is referenced by a m2m field. In that case, AlterField should be a noop, but this patch will ignore that field too as in failing test. thinking

Again, no, for exactly the same reason, check out ticket-33953. I have to double check that we can move it forward for foreign keys.

**Review by felixxm — APPROVED** [review]
@DevilsAutumn Thanks :+1: I think we can push it forward.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by DevilsAutumn** [comment]
> @DevilsAutumn Thanks :+1: I think we can push it forward.

Great!🎉

**Pull request merged**